google/cloud/internal/big_endian_test.cc

```diff
@@ -13,67 +13,179 @@
 // limitations under the License.
 
 #include "google/cloud/internal/big_endian.h"
-#include <gmock/gmock.h>
-#include <cstring>
+#include "google/cloud/testing_util/assert_ok.h"
+#include <ios>
+#include <limits>
+#include <string>
+#include <vector>
 
 namespace google {
 namespace cloud {
 inline namespace GOOGLE_CLOUD_CPP_NS {
 namespace internal {
 namespace {
 
-TEST(BigEndianTest, Int16) {
-  std::uint8_t buf[] = {0x01, 0x02};
-  std::int16_t value;
-  static_assert(sizeof(value) == sizeof(buf), "Mismatched sizes");
-  std::memcpy(&value, buf, sizeof(buf));
-  EXPECT_EQ(0x0102, FromBigEndian(value));
-  EXPECT_EQ(value, FromBigEndian(ToBigEndian(value)));
+template <typename T>
+struct TestData {
+  T n;
+  std::string s;
+};
+
+template <typename T>
+void RunTests(std::vector<TestData<T>> const& test_data) {
+  for (auto const& td : test_data) {
+    std::string const encoded = EncodeBigEndian(td.n);
+    // The std::hex io manipulator doesn't work right with `char`, so we add 0
+    // (an int) to it so that small types are promoted to at least an int.
+    auto const for_hex = td.n + 0;
```

Comment:
Instead of doing this, why not be explicit with `static_cast<int>`?

Reply:
If the type of `td.n` was, say, int64_t, then casting to int would lose information. Using `+ 0` uses the normal integer promotion rules and will promote values smaller than int to int, but types larger than int will stay the same.